### PR TITLE
Collapsed forwarding support in ATS core

### PR DESCRIPTION
**Pull request by ezelkow1**
This adds a new write fail action, CACHE_WL_FAIL_ACTION_COLLAPSED_FORWARDING. If it fails to obtain a write lock then it will loop for open_write_retries*open_read_retry_time, each time doing a DecideCacheLookup.  In the end if it fails to ever either obtain a
write lock or find the object to read then it does the default write fail action of going to the origin.

In testing using duplicated requests I can see multiple simultaneous requests end up in this read loop and eventually get a hit in cache while the original request gets written.  Occaisonally there is still an extra hit or two to the origin which I believe is
state machine dependant, so sort of unavoidable, sometimes you can have the original request finish the write while a duplicated request ends up obtaining the write lock directly afterwards and so it hits the origin as well. This may be unavoidable unless cache
lookups are done everywhere before a write.

On an origin fail it will do the default of sending back an ERR_CONNECT_FAIL

### CODE REVIEW

File: proxy/http/HttpCacheSM.cc [hunk 1]
```diff
@@ -196,10 +196,17 @@ HttpCacheSM::state_cache_open_write(int event, void *data)
           "retrying cache open write...",
           master_sm->sm_id, open_write_tries);
 
-    open_write(
-      &cache_key, lookup_url, read_request_hdr, master_sm->t_state.cache_info.object_read,
-      (time_t)((master_sm->t_state.cache_control.pin_in_cache_for < 0) ? 0 : master_sm->t_state.cache_control.pin_in_cache_for),
-      retry_write, false);
+    if (master_sm->t_state.txn_conf->cache_open_write_fail_action != HttpTransact::CACHE_WL_FAIL_ACTION_COLLAPSED_FORWARDING) {
```
Comment (sudheerv) [inline proxy/http/HttpCacheSM.cc]: case CACHE_EVENT_OPEN_WRITE_FAILED:
    // if (master_sm->t_state.txn_conf->cache_open_write_fail_action == //CACHE_OPEN_WRITE_FAIL_ACTION::READ_RETRY) {
// just fall back to open_read_retry settings
//open_write_tries = master_sm->t_state.txn_conf->max_cache_open_write_retries;
 //}

    if (open_write_tries <= master_sm->t_state.txn_conf->max_cache_open_write_retries) {
      // Retry open write;
      open_write_cb = false;
      do_schedule_in();
    } else {

File: proxy/http/HttpCacheSM.cc [hunk 2]
```diff
@@ -196,10 +196,17 @@ HttpCacheSM::state_cache_open_write(int event, void *data)
...
+      open_write(
+        &cache_key, lookup_url, read_request_hdr, master_sm->t_state.cache_info.object_read,
+        (time_t)((master_sm->t_state.cache_control.pin_in_cache_for < 0) ? 0 : master_sm->t_state.cache_control.pin_in_cache_for),
+        retry_write, false);
+    } else {
+      Debug("http_cache", "[%" PRId64 "] [state_cache_open_write] cache open write failure INTERVAL. Saw CF, not attempting write",
+            master_sm->sm_id);
+      open_write_cb = false;
+      master_sm->handleEvent(CACHE_EVENT_OPEN_WRITE_FAILED, data);
```
Comment (sudheerv) [inline proxy/http/HttpCacheSM.cc]: change this to 
// the idea is to go back to read state on write fail (as opposed to relying on  write_fail action)
// master_sm->handleEvent(CACHE_EVENT_OPEN_READ, data)

**Comment by bryancall** [comment]
[approve ci]

**Comment by ezelkow1** [comment]
superseded by https://github.com/apache/trafficserver/pull/6028 since @sudheerv had a chance to work on it more

**Pull request closed without merging**